contracts/modules/v2/StakingPool.sol

```diff
@@ -507,6 +507,131 @@ contract StakingPool is IStakingPool, ERC721 {
+  /// @param newTrancheId      The id of the new tranche determining the new deposit period.
+  /// @param topUpAmount       An optional amount if the user wants to also increase the deposit
+  ///                          amount.
+  function extendDeposit(
+    uint tokenId,
+    uint initialTrancheId,
+    uint newTrancheId,
+    uint topUpAmount
+  ) public {
+    uint _firstActiveTrancheId = block.timestamp / TRANCHE_DURATION;
+    uint maxTranche = _firstActiveTrancheId + MAX_ACTIVE_TRANCHES;
+    require(tokenId != 0, "StakingPool: Invalid token id");
+    require(
+      _isApprovedOrOwner(msg.sender, tokenId),
+      "StakingPool: Not authorized to extend deposits on this token"
+    );
+    require(
+      initialTrancheId < newTrancheId,
+      "StakingPool: The chosen tranche cannot end before the initial one"
+    );
+    require(newTrancheId <= maxTranche, "StakingPool: The chosen tranche is not available yet");
+    require(
+      newTrancheId >= _firstActiveTrancheId,
+      "StakingPool: The chosen tranche has already reached the maturity date"
+    );
+
+    // If the intial tranche is expired, withdraw everything and make a new deposit equal to the
+    // withdrawn stake amount plus a top up amount if applicable. This will require the user to
+    // grant sufficient allowance beforehand.
+    if (initialTrancheId < _firstActiveTrancheId) {
+      uint[] memory trancheIds = new uint[](1);
+      trancheIds[0] = initialTrancheId;
+
+      WithdrawParams[] memory withdrawParams = new WithdrawParams[](1);
+      withdrawParams[0] = WithdrawParams(
+        tokenId,
+        true, // Withdraw deposit
+        true, // Withdraw rewards
+        trancheIds
+      );
+
+      (uint withdrawnStake, /* uint rewardsToWithdraw */) = withdraw(withdrawParams);
+
+      depositTo(
+        withdrawnStake + topUpAmount,
+        newTrancheId,
+        tokenId,
+        msg.sender
+      );
+
+      return; // Done! Skip the rest of the function.
+    }
+
+    // If the initial tranche is still active, move all the shares and pending rewards to the
+    // newly chosen tranche and its coresponding deopsit.
+
+    // First make sure tranches are up to date in terms of accumulated NXM rewards.
+    updateTranches();
+
+    Tranche memory initialTranche = tranches[initialTrancheId];
+    Tranche memory newTranche = tranches[newTrancheId];
+    Deposit memory initialDeposit = deposits[tokenId][initialTrancheId];
+    Deposit memory newDeposit = deposits[tokenId][newTrancheId];
+    uint _accNxmPerRewardsShare = accNxmPerRewardsShare;
+
+
+    // The user's shares are moved from the initial tranche to the new one.
+    initialTranche.stakeShares -= initialDeposit.stakeShares;
+    initialTranche.rewardsShares -= initialDeposit.rewardsShares;
+    newTranche.stakeShares += initialDeposit.stakeShares;
+    newTranche.rewardsShares += initialDeposit.rewardsShares;
```

Comment:
We're extending the period so we should increase the reward shares the user gets for this deposit, the formula is inside the `calculateRewardSharesAmount` function (not confirmed to be the final version though).

Reply:
You could actually pick up the `calculateRewardSharesAmount` task, it's quite contained and requires some brainstorming and checks. See NM-514

Reply:
Additionally, when minting new reward shares you should also mint some reward shares for the pool fee as well, see [this commit from PR 219](https://github.com/NexusMutual/smart-contracts/pull/219/commits/31225054e1bacfefe2cf690209df59df30286995)